profiler/src/Profiler.cc

```diff
@@ -115,3 +115,10 @@ bool Profiler::SetImplementation(std::unique_ptr<ProfilerImpl> _impl)
   gzdbg << "Gazebo profiling with: " << this->impl->Name() << std::endl;
   return true;
 }
+
+//////////////////////////////////////////////////
+Profiler *Profiler::Instance()
+{
+  static Profiler profiler;
+  return &profiler;
```

Comment:
```suggestion
  return SingletonT<Profiler>::Instance();
```

I think doing this might maintain ABI compatibility. Can you give it a try?

Reply:
And about NeverDestroyed (very short version). IMHO it's a very dirty hack.
1. Valgrind will see indirect memory leak if object inside it allocates memory.
2. AFAIU it was invented for cases of library loading with dlopen and unloaded before termination. In this case code of destructor of static object is unloaded and attempt to call it will lead to crash on termination (I saw similar case with destructor of exception on plugin loading mechanism in one of project where I worked)
3. I guess it can be treated as UB. Can deep dive in standard to find more accurate interpretation, but it will take a lot of time

Reply:
I think you're right that it'll probably work fine during initialization, but your approach would still be problematic during destruction since the class is not trivially destructible. We follow the Google C++ style guide and if you look at the "Common patterns" subsection under [Static and Global Variables](
https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables), you will see that the preferred approach is to do `static const auto& impl = *new T(args...);`, which is what `NeverDestroyed` does.

My preference would be that we would consistently use `NeverDestroyed` as a pattern for singletons as it's already used in other places in the code base and complies with the Google style guide.